lib/plugins/helpers/network.rb

```diff
@@ -1,49 +1,76 @@
-class Plugin
-  def bot_ip
-    option = @config['managed'] ? 'private' : 'public' if @config['type'] == 'api'
-    option = 'private' unless @config['type'] == 'api'
+class Network
+  def bot_ip(config)
+    option = config['managed'] ? 'private' : 'public' if config['type'] == 'api'
+    option = 'private' unless config['type'] == 'api'
 
     local_ip = Socket.ip_address_list.detect(&:ipv4_private?).ip_address
     public_ip = Socket.ip_address_list.detect(&:ipv4?).ip_address
-    @bot_ip = option == 'public' ? public_ip : local_ip
+    option == 'public' ? public_ip : local_ip
   end
 
-  def plugin_ip
-    container_info = Docker::Container.get(@name).info
+  def plugin_ip(logger, config, name)
+    container_info = Docker::Container.get(name).info
     # Builds URL based on Container Settings
     container_ip = container_info['NetworkSettings']['IPAddress'].to_s
     local_ip = Socket.ip_address_list.detect(&:ipv4_private?).ip_address
 
     # Determine if running via DIND/Compose Config or if running local
     compose_bot = local_ip.rpartition('.')[0] == container_ip.rpartition('.')[0]
-    @logger.debug(compose_bot ? "Plugin: #{@name}: running inside DIND" : "Plugin: #{@name}: running on local machine")
+    logger.debug(compose_bot ? "Plugin: #{name}: running inside DIND" : "Plugin: #{name}: running on local machine")
 
-    exposed_ip = @config['public'] ? local_ip : '127.0.0.1'
+    exposed_ip = config['public'] ? local_ip : '127.0.0.1'
     # If Compose, set docker network ip. If, local use system IP set above
-    @plugin_ip = compose_bot ? container_ip : exposed_ip
+    compose_bot ? container_ip : exposed_ip
   end
 
-  def expose
-    expose_ip = @config['public'] ? '0.0.0.0' : '127.0.0.1'
-    container_port = @container_hash[:ExposedPorts].keys[0].to_s
-    @config['app_port'] = container_port.chomp('/tcp') if @config['app_port'].blank?
+  def expose(config)
+    expose_ip = config['public'] ? '0.0.0.0' : '127.0.0.1'
     exposed_port = {
       HostConfig: {
         PortBindings: {
-          "#{@config['app_port']}/tcp" =>
+          "#{config['app_port']}/tcp" =>
           [{
             HostIp: expose_ip.to_s,
-            HostPort: @config['exposed_port'].to_s
+            HostPort: config['exposed_port'].to_s
           }]
         }
       }
     }
-    @container_hash.merge!(exposed_port)
+    exposed_port
+  end
+
+  def port_find(dynamic_port)
+    port_check = port_available?(dynamic_port)
+    unless port_check
+      dynamic_port += 1
+      while port_check == false
+        port_check = port_available?(dynamic_port)
+        break if port_check
+        dynamic_port += 1
+      end
+    end
+    dynamic_port
+  end
+
+  def port_available?(port)
+    Timeout.timeout(1) do
+      begin
+        TCPSocket.new('127.0.0.1', port).close
+        false
+      rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH, SocketError
+        true
+      end
+    end
+  rescue Timeout::Error
```

Comment:
With the nested begin-rescue-end i find this less readable to leave off the initial begin.  Maybe this was a recommendation by Rubocop?  If so breaking the internal begin-rescue-end into another function would likely be more readable.